### PR TITLE
(CDAP-11411) Move all Spark container launching code into cdap-spark-core-base

### PR DESCRIPTION
**Pull request by chtyim**
This PR is the first step for CDAP-11411, speed up of Spark launch. It main do refactoring so that there is no more spark specific code in cdap-app-fabric. Also it simplifies the code structural by only having `DistributedXXXProgramRunner` for running a specific program type, removing the need of program specific `XXXProgramTwillApplication`.

### CODE REVIEW

**Comment by chtyim** [comment]
Testing on cluster which can launch and run different program types successfully.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java [hunk 1]
```diff
@@ -198,56 +162,92 @@ protected void validateOptions(Program program, ProgramOptions options) {
...
+   *         be started in the same order
+   */
+  protected Iterable<Set<String>> getRunnableLaunchOrder(Map<String, RunnableResource> runnables) {
+    return Collections.emptyList();
+  }
+
+  /**
+   * Returns the set of extra resources that need to be localized to Twill containers.
+   *
+   * @param program te program to be launched in Twill container.
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java]: typo: te

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by chtyim** [comment]
Build passed https://builds.cask.co/browse/CDAP-DUT5739

**Pull request merged**